README.md
Repository for the development version of water package.

[![](http://cranlogs.r-pkg.org/badges/grand-total/water)](https://cran.r-project.org/package=water)


water package is available at [CRAN network](https://cran.r-project.org/package=water)

You can install development version with: 


    install.packages("devtools")
    library(devtools)
    install_github("midraed/water")


#Package: water#
Title: Actual Evapotranspiration with Surface Energy Balance models

Author: Guillermo Federico Olmedo, Samuel Ortega-Farías, David Fonseca-Luengo,
  Daniel de la Fuente-Saiz and Fernando Fuentes-Peñailillo

Maintainer: Guillermo Federico Olmedo <guillermo.olmedo@gmail.com>

Description: Tools and functions to calculate actual evapotranspiration using surface energy balance models. 
Depends: R (>= 3.2.1), raster (>= 2.1), sp (>= 1.1-1), proj4, stringr

License: GPL (>= 2)

## Contributing

Contributions are accepted by [fork](https://help.github.com/articles/fork-a-repo/) and [pull-request](https://help.github.com/articles/using-pull-requests/)




NEWS
water: 
change log. From 09/09/2015. Most recent changes first / on top.
Guillermo OLMEDO
-----------------------------------------------------------------
VERSION 0.8.4

* Fixed read.WSdata example using a fixed (the correct one) timezone. Before this
  timezone of the example was estimated with user timezone
* Fixed mistmatch between LAI code and LAI documentation
* Suppress confusing warning on read.WSdata()


VERSION 0.8

* New method for anchors pixels selection named "flexible". This method
  will find the pixel with the best condition when there aren't any perfect 
  candidates. It will relax the conditions needed and then adjust the ETp.coef
  based on Gonzalez et al 2015. See calcH()
* Added a new method to G estimation. Now the method can be choosen from 
  soilHeatFlux() and also in METRIC.EB()

VERSION 0.7

* New S3 class for the results of METRIC.EB(). The new class includes the
  RasterStack with the LSEB results, the weather station data used, and the
  parameters and methods selected in METRIC.EB().
* When using METRIC.EB(), calcAnchors always uses method="short.crops" not 
  matter the method selected for Zom.method.
* water can handle filenames for Landsat images using the pre-collection or
  the collection 1 sets.
* New method for calcAnchors: "flexible". This method is based in the "best" 
  method, but, will relax the requesites for the anchors when there is no pixel
  with those conditions. "flexible" is the default method now.
* in LAI() the method "vineyard" used the toa reflectancy image. This was 
  corrected and now uses the radiance image
* Added a function to estimate radiance from a digital counts image and a 
  metadata file: calcRadiance().
* Several minor bugs fixed.


VERSION 0.6.1

* Added a workaround for read.WSdata() when there is not rain data
* Fixed a bug in momentumRoughness() when using the Perrier equation
* Added extraParameters to METRIC.EB. This allows to pass more parameters to 
  the internal functions and apply different methods. i.e. extra parameters for
  different Zom.methods.

VERSION 0.6

* Added a new function loadImageSR to load surface reflectance from
  Landsat 8 from ESPA files
* calcSR only works for L7
* calcTOAr() now estimates TOA reflectance for Landsat 8
* surfaceTemperature() uses the scene in DN and estimates the brightness
  temperature for Landsat 7 or 8. Argument thermalband is deprecated.
* Added a new example of weather station for the new Landsat 8 example
* Added a Landsa 8 scene as example dataset in raw format complete with the
  surface reflectance estimated by ESPA
* Added a new vignette describing how to use Landsat 8 data in water package

VERSION 0.5.1

* Fixed checkSRTMgrids for positives values of lat/long.
* Small improvement in waterWeatherStation plot. Now the time axis is always 
  complete.
* Added daily summary for WeatherStation objects
* Fixed minor bug in vignette

VERSION 0.5

* Added maxit parameter to calcH to control the maximun number of iterations.
* Added an optional constrain to the selection of anchors pixels using the 
  location of the weather station and a 30km buffer.
* Changed default value for Z.om.ws in calcH. (From 0.0018 to 0.03)
* Added a new parameter to calcAnchors, available for both methods: buffer.
  buffer allow to set the minimun distance between two anchor pixels of the same
  kind
* Added a new method for calcAnchors = "CITRA-MCBbc". This method chooses the 
  coldest and hottest anchors pixels availables. Previous method ("CITRA-MCBr")
  chooses random pixel who meets the conditions. CITRA-MCBbc is now the default
  method for calcAnchors
* General remote sensing functions moved to a separate file
* loadImage detects when there is more than 1 image on the working folder
* Rn, G, H, LE are restricted to values > 0
* Improvements to anchors pixels selection: more releaxed hot 
  temperature criterium, distance, mean of many pixels, etc
* Added two methods for land surface estimation: single channel and split windows.
  Split windows only works for Landsat 8.
* loadImage now loads thermal data also: low gain for L7 and both bands for L8
* Fixed big bug when estimating ETo with a large weather station file

VERSION 0.3.2

* Solved errors when using non ASCII characters
* Fixes some bug who prevent using water with L8 data
* More flexibility on band names and changes to reflect the changes on USGS's 
  ESPA file naming
* Improvement to plot.waterWeatherStation to plot only data from the sattelite
  overpass when the object was created using a MTL file.

VERSION 0.3

* First release to CRAN!

VERSION 0.2.9

* Added dailyET function to calc and sum hourly ET for a complete day
* Added new parameter to hourlyET and dailyET to choose between ETo or ETr
* read.WSdata doesn't require a Landsat metadata anymore. Now you can process
  weatherstation data without a landsat flyby. print method changed to reflect
  this
* function METRIC.EB now accepts many parameters for method of the included 
  functions. For example, you can set with method you choose for LAI estimation
* We're approaching CRAN launch date. Changed version numbering to reflect this. 
  First CRAN version will be 0.3

VERSION 0.0.2.4

* Added function to prepare data from Weather Station and calc weather
  conditions at satellite flyby.
* Added function to calculate Surface Energy Balance using METRIC

VERSION 0.0.2.3

* Rename sensibleHeatFlux() to sensibleHeatFlux_old()
* Remove sensibleHeatFlux.CITRA()
* Split sensibleHeatFlux in 2 new functions:
  - calcAnchors() for anchors points
  - calcH() for the iterative process to calculate H and dT
* Added function to estimate Net Radiation


R/water_METRIC.R
#' Estimates Net Radiation as in METRIC Model
#' @param image.DN         raw imagen in digital counts to evaluate
#' @param WeatherStation   Weather Station data, can be a waterWeatherStation 
#' object
#' @param MTL              Landsat metadata file
#' @param sat              Landsat satellite version. "L7" or "L8"
#' @param thermalband      Landsat low gain thermalband
#' @param plain            Logical. If TRUE surface is assumed plain
#' @param DEM              Digital Elevation Model of the study area. Not needed
#' if plain = TRUE
#' @param aoi              SpatialPolygon object with limits of Area of interest
#' @param alb.coeff        coefficient to transform narrow to broad band albedo.
#' See Details.
#' @param LAI.method       Method used to estimate LAI from spectral data. 
#' See Details.
#' @family METRIC model functions
#' @author Guillermo F Olmedo, \email{guillermo.olmedo@@gmail.com}
#' @references
#' R. G. Allen, M. Tasumi, and R. Trezza, "Satellite-based energy balance for mapping evapotranspiration with internalized calibration (METRIC) - Model" Journal of Irrigation and Drainage Engineering, vol. 133, p. 380, 2007
#' @export
METRIC.Rn <- function(image.DN, WeatherStation, MTL, sat = "auto", thermalband, 
                      alb.coeff = "Tasumi", LAI.method = "metric2010", 
                      plain = TRUE, DEM, aoi){
  path=getwd()
  #pb <- txtProgressBar(min = 0, max = 100, style = 3)
  if(plain==TRUE){
    DEM <- raster(image.DN[[1]])
    values(DEM) <- WeatherStation$location$elev
  }
  surface.model <-METRICtopo(DEM)
  #setTxtProgressBar(pb, 3)
  solar.angles.r <- solarAngles(surface.model = surface.model, 
                                WeatherStation = WeatherStation, MTL = MTL)
  Rs.inc <- incSWradiation(surface.model = surface.model, 
                           solar.angles = solar.angles.r, 
                           WeatherStation = WeatherStation)
  if(sat=="L7" | sat=="L8"){
  image.TOAr <- calcTOAr(image.DN = image.DN, sat=sat, MTL = MTL, 
                         incidence.rel = solar.angles.r$incidence.rel)
  image.SR <- calcSR(image.TOAr=image.TOAr, sat = sat, 
                     surface.model=surface.model, 
                     incidence.hor = solar.angles.r$incidence.hor, 
                     WeatherStation=WeatherStation)}
  if(sat=="MODIS"){image.SR <- image.DN}
  albedo <- albedo(image.SR = image.SR,  coeff=alb.coeff, sat=sat)
  #setTxtProgressBar(pb, 6)
  if(sat=="MODIS"){image.TOAr <- image.DN} # Only used for LAI estimation,
                                           # and some LAI models, use SR
  LAI <- LAI(method = LAI.method, image = image.TOAr, L=0.1)
  if(sat=="L7" | sat=="L8"){
  Ts <- surfaceTemperature(LAI=LAI, sat = sat, thermalband = thermalband,
                           WeatherStation = WeatherStation)}
  if(sat=="MODIS"){Ts <- image.DN$LST}
  #setTxtProgressBar(pb, 35)
  Rl.out <- outLWradiation(LAI = LAI, Ts=Ts)
  Rl.inc <- incLWradiation(WeatherStation,DEM = surface.model$DEM, 
                           solar.angles = solar.angles.r, Ts= Ts)
  Rn <- netRadiation(LAI, albedo, Rs.inc, Rl.inc, Rl.out)
  plot(Rn, main="METRIC Net Radiation")
  return(Rn)
}

#' Estimates Net Radiation as in METRIC Model
#' @param image.DN         raw imagen in digital counts to evaluate
#' @param WeatherStation   Weather Station data, can be a waterWeatherStation 
#' object
#' @param Rn               RasterLayer with Net Radiation data in W/m2
#' @param plain            Logical. If TRUE surface is assumed plain
#' @param DEM              Digital Elevation Model of the study area. Not needed
#' if plain = TRUE
#' @param aoi              SpatialPolygon object with limits of Area of interest
#' @author Guillermo F Olmedo, \email{guillermo.olmedo@@gmail.com}
#' @family METRIC model functions
#' @references 
#' R. G. Allen, M. Tasumi, and R. Trezza, "Satellite-based energy balance for mapping evapotranspiration with internalized calibration (METRIC) - Model" Journal of Irrigation and Drainage Engineering, vol. 133, p. 380, 2007
#' @export
METRIC.G <- function(image.DN, WeatherStation=WeatherStation, Rn,  
                     plain = TRUE, DEM, aoi){
  path=getwd()
  if(plain==TRUE){
    DEM <- raster(image.DN[[1]])
    values(DEM) <- WeatherStation$location$elev
  } 
  surface.model <-METRICtopo(DEM)
  solar.angles.r <- solarAngles(surface.model = surface.model)
  image.TOAr <- calcTOAr(image.DN = image.DN, 
                         incidence.rel = solar.angles.r$incidence.rel)
  image.SR <- calcSR(image.TOAr=image.TOAr, 
                      surface.model=surface.model, 
                      incidence.hor = solar.angles.r$incidence.hor, 
                      WeatherStation=WeatherStation, sat="auto")
  albedo <- albedo(image.SR = image.SR)
  Ts <- surfaceTemperature(sat = "auto" )
  G <- soilHeatFlux(image = image.SR, Ts=Ts,albedo=albedo, Rn)
}


#' Estimates Energy Balance using METRIC2010 Model
#' @param image.DN         raw imagen in digital counts to evaluate
#' @param image.SR         L8 ONLY. Surface reflectance imagen. water package does not 
#' include a model to calculate surface reflectance for Landsat 8 images. Landsat 8 users 
#' should download precalculated surface reflectances from espa website 
#' (espa.cr.usgs.gov). 
#' @param WeatherStation   Weather Station data, can be a waterWeatherStation 
#' object
#' @param MTL              Landsat metadata file
#' @param sat              Landsat satellite version. "L7" or "L8"
#' @param thermalband      Landsat low gain thermalband
#' @param plain            Logical. If TRUE surface is assumed plain
#' @param DEM              Digital Elevation Model of the study area. Not needed
#' if plain = TRUE
#' @param aoi              SpatialPolygon object with limits of Area of interest
#' @param G.method         method used for the G estimation. Currently implemeted are 
#'                         "Tasumi" for Tasumi,2003 or "Bastiaanssen" for Bastiaanssen, 2000

#' @param alb.coeff        coefficient to transform narrow to broad band albedo.
#' See Details.
#' @param LST.method       Method for land surface temperature estimation. "SC" 
#' for single channel or "SW" for split window algorithm. 
#' "SW" is only available for L8. See \code{water::surfaceTemperature}
#' @param LAI.method       Method used to estimate LAI from spectral data. 
#' See Details.
#' @param L                L value for SAVI calculation
#' @param Zom.method       method selected to calculate momentum roughness 
#' length. Use "short.crops" for short crops methods from Allen et al (2007); 
#' "custom" for custom method also in Allen et al (2007); Or "Perrier" to use 
#' Perrier equation as in Santos et al (2012) and Pocas et al (2014).
#' @param anchors.method   method for the automatic selection of the anchor pixels. 
#' @param anchors          data.frame or SpatialPointsDataFrame with the anchor
#'                         pixels. The data frame must include a "type" column 
#'                         with "hot" and "cold" values.
#' @param n                number of pair of anchors pixels to calculate
#' @param ETp.coef         ETp coefficient usually 1.05 or 1.2 for alfalfa
#' @param Z.om.ws          momentum roughness lenght for WeatherStation. Usually
#' 0.0018 or 0.03 for long grass
#' @param verbose          Logical. If TRUE will print aditional data to console
#' @param extraParameters  Extra parameters for the non default methods. i.e. 
#' Zom.method = "Perrier", needs two extra parameters: fLAI, h. See 
#' help(momentumRoughnessLength).
#' @details
#' There are differents models to convert narrowband data to broadband albedo. 
#' You can choose alb.coeff ="Tasumi" to use Tasumi et al (2008) coefficients, 
#' calculated for Landsat 7; alb.coeff ="Liang" to use Liang Landsat 7 
#' coefficients or "Olmedo" to use Olmedo coefficients for Landsat 8.
#' @section Extra Parameters:
#' Extra Paramenters for functions inside METRIC.EB() include:
#' * for momentumRoughness when Zom.method = "Perrier": fLAI, h.
#' * for calcAnchors(): minDist, WSbuffer, deltaTemp
#' @author Guillermo F Olmedo, \email{guillermo.olmedo@@gmail.com}
#' @family METRIC model functions
#' @references 
#' R. G. Allen, M. Tasumi, and R. Trezza, "Satellite-based energy balance for mapping evapotranspiration with internalized calibration (METRIC) - Model" Journal of Irrigation and Drainage Engineering, vol. 133, p. 380, 2007
#' González, Arturo & Hay, Christopher & Kjaersgaard, Jeppe & Neale, Christopher. (2015). Use of Remote Sensing to Generate Crop Coefficient and Estimate Actual Crop Evapotranspiration. 10.13031/aim.20152190105. 
#' @examples 
#' ### Data preparation
#' library(water)
#' aoi <- createAoi(topleft = c(500000, -3644000), bottomright = c(526000, -3660000))
#' raw_data_folder <- system.file("extdata", package="water")
#' image <- loadImage(path=raw_data_folder, aoi=aoi, sat="L8")
#' image.SR <- loadImageSR(path=raw_data_folder, aoi=aoi)
#' csvfile <- system.file("extdata", "INTA.csv", package="water")
#' MTLfile <- system.file("extdata", "LC82320832016040LGN00_MTL.txt", package="water")
#' \dontrun{
#' WeatherStation <- read.WSdata(WSdata = csvfile, 
#'                               datetime.format =  "%Y/%m/%d %H:%M", 
#'                               columns = c("datetime", "temp",
#'                                           "RH", "pp", "radiation", "wind"), 
#'                               lat=-33.00513, long= -68.86469, elev=927, height= 2,
#'                               MTL=MTLfile)
#'                               
#' ### LSEB with default methods and no extra parameters                               
#' Energy.Balance <- METRIC.EB(image.DN = image, image.SR = image.SR,
#'                             plain=TRUE, aoi=aoi, n = 5, WeatherStation = WeatherStation, 
#'                             ETp.coef = 1.2, sat="L8", alb.coeff = "Olmedo", LST.method = "SW", 
#'                             LAI.method = "metric2010", Z.om.ws = 0.03, MTL = MTLfile)
#'                             
#'                             
#' ### LSEB with "Perrier" method for Zom and extra parameters                               
#' Energy.Balance <- METRIC.EB(image.DN = image, image.SR = image.SR,
#'                             plain=TRUE, aoi=aoi, n = 5, WeatherStation = WeatherStation, 
#'                             ETp.coef = 1.2, sat="L8", alb.coeff = "Olmedo", LST.method = "SW", 
#'                             LAI.method = "metric2010", Zom.method = "Perrier", Z.om.ws = 0.03, 
#'                             MTL = MTLfile, extraParameters = c(fLAI = 0.5, h = 1.8) ) 
#' }
#' @export
METRIC.EB <- function(image.DN, image.SR, WeatherStation, MTL, sat = "auto",
                      thermalband, plain=TRUE, DEM, aoi, G.method = "Tasumi",
                      alb.coeff = "Tasumi", LST.method = "SC",
                      LAI.method = "metric2010", L = 0.1,
                      Zom.method = "short.crops", anchors.method = "CITRA-MCB",
                      anchors, n = 1, ETp.coef= 1.05, Z.om.ws=0.0018,
                      verbose = FALSE, extraParameters = vector()){
  path=getwd()
  #pb <- txtProgressBar(min = 0, max = 100, style = 3)
  if(plain==TRUE){
    DEM <- raster(image.DN[[1]])
    values(DEM) <- WeatherStation$location$elev
  }
  surface.model <-METRICtopo(DEM)
  #setTxtProgressBar(pb, 3)
  if(missing(MTL)){MTL <- list.files(path = path, pattern = "MTL.txt", full.names = T)}
  solar.angles.r <- solarAngles(surface.model = surface.model, 
                                WeatherStation = WeatherStation, MTL = MTL)
  Rs.inc <- incSWradiation(surface.model = surface.model, 
                           solar.angles = solar.angles.r, 
                           WeatherStation = WeatherStation)
  if(sat=="L7" | sat=="L8"){
    image.TOAr <- calcTOAr(image.DN = image.DN, sat=sat, MTL = MTL,
                           incidence.rel = solar.angles.r$incidence.rel)
    if(sat=="L7"){
      image.SR <- calcSR(image.TOAr=image.TOAr, sat = sat, 
                         surface.model=surface.model, 
                         incidence.hor = solar.angles.r$incidence.hor, 
                         WeatherStation=WeatherStation)}
    }
  if(sat=="MODIS"){image.SR <- image.DN}
  albedo <- albedo(image.SR = image.SR,  coeff=alb.coeff, sat=sat)
  #setTxtProgressBar(pb, 6)
  if(sat=="MODIS"){image.TOAr <- image.DN} # Only used for LAI estimation,
  # and some LAI models, use SR
  if(LAI.method == "vineyard"){
    image.rad <- calcRadiance(image.DN, sat = sat, MTL = MTL)
    LAI <- LAI(method = LAI.method, image = image.rad, L=L)
    }
  if(LAI.method == "turner"){LAI <- LAI(method = LAI.method, image = image.SR, L=L)}
  if (LAI.method != "vineyard" & LAI.method != "turner"){LAI <- LAI(method = LAI.method, image = image.TOAr, L=L)}
  if(sat=="L7" | sat=="L8"){
    Ts <- surfaceTemperature(LAI=LAI, sat = sat, image.DN=image.DN,
                             WeatherStation = WeatherStation, method = LST.method)}
  if(sat=="MODIS"){Ts <- image.DN$LST}
  #setTxtProgressBar(pb, 35)
  Rl.out <- outLWradiation(LAI = LAI, Ts=Ts)
  Rl.inc <- incLWradiation(WeatherStation,DEM = surface.model$DEM, 
                           solar.angles = solar.angles.r, Ts= Ts)
  Rn <- netRadiation(LAI, albedo, Rs.inc, Rl.inc, Rl.out)
  # Rn[Rn < 0]  <-  0 # see H
  #setTxtProgressBar(pb, 40)
  G <- soilHeatFlux(image = image.SR, Ts=Ts,albedo=albedo, 
                    Rn=Rn, LAI=LAI, method = G.method)
  # G[G < 0]  <-  0 # see H
  if(Zom.method == "short.crops"){
    Z.om <- momentumRoughnessLength(LAI=LAI, mountainous = !plain, 
                                    method = Zom.method, 
                                    surface.model = surface.model)
  }
  if(Zom.method == "custom"){
    Z.om <- momentumRoughnessLength(LAI=LAI, mountainous = !plain, 
                                    method = Zom.method, a = extraParameters["a"],
                                    b = extraParameters["b"],
                                    surface.model = surface.model)
  }
  if(Zom.method == "Perrier"){
    Z.om <- momentumRoughnessLength(LAI=LAI, mountainous = !plain, 
                                    method = Zom.method, fLAI = extraParameters["fLAI"],
                                    h = extraParameters["h"],
                                    surface.model = surface.model)
  }
  if(Zom.method != "short.crops"){Z.om.sc <- momentumRoughnessLength(LAI=LAI, mountainous = !plain, 
                                                             method = "short.crops", 
                                                             surface.model = surface.model)
  } else {Z.om.sc <- Z.om}
  par(mfrow=c(1,2))
  if(missing(anchors)){
    if(is.na(extraParameters["deltaTemp"])){extraParameters['deltaTemp'] = 5}
    if(is.na(extraParameters["minDist"])){extraParameters['minDist'] = 500}
    if(is.na(extraParameters["WSbuffer"])){extraParameters['WSbuffer'] = 30000}
    anchors <- calcAnchors(image = image.TOAr, Ts = Ts, LAI = LAI, plots = T,
                           albedo = albedo, Z.om = Z.om.sc, n = n, 
                           anchors.method = anchors.method, WeatherStation = WeatherStation,
                           deltaTemp = extraParameters['deltaTemp'],
                           minDist = extraParameters['minDist'],
                           WSbuffer = extraParameters['WSbuffer'],
                           verbose = verbose)
    print(anchors)
  }
  #setTxtProgressBar(pb, 45)
  if(anchors.method == "flexible"){
     flex.cold <- anchors$flex[anchors$type == "cold"][1]
     NDVIcold <- mean(as.numeric(anchors$NDVI[anchors$type == "cold"]))
     if(flex.cold > 0 & NDVIcold < 0.8){
       print(paste0("mean NDVI for cold pixel: ", NDVIcold))
       # NDVI and Kc relationship from:
       #González, Arturo & Hay, Christopher & Kjaersgaard, Jeppe & Neale, 
       #Christopher. (2015). Use of Remote Sensing to Generate Crop Coefficient 
       #and Estimate Actual Crop Evapotranspiration. 10.13031/aim.20152190105. 
       ETp.coef <-  2.11 * NDVIcold - 0.4989 
       print(paste0("ETp.coef updated to: ", ETp.coef, " (Gonzalez et al, 2015)"))
     }
  }
  H <- calcH(anchors = anchors, Ts = Ts, Z.om = Z.om, mountainous = !plain,
             WeatherStation = WeatherStation, ETp.coef = ETp.coef,
             Z.om.ws = Z.om.ws, DEM = DEM, Rn = Rn, G = G, verbose = verbose)
  par(mfrow=c(1,1))
  #setTxtProgressBar(pb, 99)
  H <-  H$H
  # H[H < 0]  <-  0  not a good solution...! not for any of the components
  LE <- Rn - G - H
  LE[LE < 0]  <-  0   # see H
  EB <- stack(Rn, G, H, LE, Ts)
  EB <- saveLoadClean(imagestack = EB,
                stack.names = c("NetRadiation", "SoilHeat", "SensibleHeat", 
                                "LatentHeat", "surfaceTemperature"), 
                file = "EB", overwrite=TRUE)
  #setTxtProgressBar(pb, 100)
  result <- list()
  result$EB <- EB
  result$WeatherStation <- WeatherStation
  coordinates(anchors) <- ~ X + Y
  result$anchors <- anchors
  result$methods <- c(sat = sat, alb.coeff = alb.coeff, LST.method = LST.method,
                      LAI.method = LAI.method, Zom.method = Zom.method, 
                      anchors.method = anchors.method, plain = plain,
                      ETp.coef= ETp.coef, Z.om.ws=Z.om.ws,
                      extraParameters = extraParameters)
  class(result) <- "waterLSEB"
  return(result)
}
  
